mbed_targets/_internal/target_attribute_hierarchy_parsers/overriding_attribute_parser.py

```diff
@@ -0,0 +1,95 @@
+"""Functions for parsing the inheritance for overriding attributes.
+
+Accumulating attributes are defined and can be overridden further down the hierarchy.
+The hierarchy is determined as 'depth-first' in multiple inheritance.
+eg. Look first for the attribute in the current target. If not found,
+look for the attribute in the first target's parent, then in the parent of the parent and so on.
+If not found, look for the property in the rest of the target's parents, relative to the
+current inheritance level.
+
+This means a target on a higher level could potentially override one on a lower level.
+"""
+from collections import deque
+from functools import reduce
+from typing import Dict, List, Any, Deque
+
+from mbed_targets._internal.target_attribute_hierarchy_parsers.accumulating_attribute_parser import (
+    ALL_ACCUMULATING_ATTRIBUTES,
+)
+
+
+def _targets_override_hierarchy(all_targets_data: Dict[str, Any], target_name: str) -> List[dict]:
+    """List all ancestors of a target in order of overriding inheritance (depth-first).
+
+    Using a depth-first traverse of the inheritance tree, return a list of targets in the
+    order of inheritance, starting with the target itself and finishing with its highest ancestor.
+
+    Args:
+        all_targets_data: a dictionary representation of all the data in a targets.json file
+        target_name: the name of the target we want to calculate the attributes for
+
+    Returns:
+        A list of dicts representing each target in the hierarchy.
+    """
+    override_order: List[dict] = []
+
+    still_to_visit: Deque[dict] = deque()
+    still_to_visit.appendleft(all_targets_data[target_name])
+
+    while still_to_visit:
+        current_target = still_to_visit.popleft()
+        override_order.append(current_target)
+        if "inherits" in current_target.keys():
+            for parent_target in reversed(current_target["inherits"]):
+                still_to_visit.appendleft(all_targets_data[parent_target])
+
+    return override_order
+
+
+def _determine_overwritten_attributes(override_order: List[dict]) -> Dict[str, Any]:
+    """Finds all the overrideable attributes for a target from its list of ancestors.
+
+    Merges the attributes from all the targets in the hierarchy. Starts from the highest ancestor
+    reduces down to the target itself, overriding if they define the same attribute.
+
+    Removes any accumulating attributes - they will be handled by a separate parser.
+
+    Args:
+        override_order: order of inheritance for the target, starting with the target up to its highest ancestor
+
+    Returns:
+        A dictionary containing all the overridden attributes for a target
+    """
+    target_attributes = reduce(lambda x, y: _update_dict(x, y), reversed(override_order))
+    overridden_attributes = _remove_accumulating_attributes(target_attributes)
+    return overridden_attributes
+
+
+def _update_dict(dict_1: Dict[Any, Any], dict_2: Dict[Any, Any]) -> Dict[Any, Any]:
```

Comment:
I'd prefer this was called `merge_dicts` or something as the name `update_dict` makes me think we're updating the same dictionary object rather than returning a new dict.